server/src/main/java/org/opensearch/transport/Header.java

```diff
@@ -55,6 +55,7 @@ public class Header {
 
     private static final String RESPONSE_NAME = "NO_ACTION_NAME_FOR_RESPONSES";
 
+    private final TransportProtocol protocol;
```

Comment:
I like this approach better but AFAIR, it brings issues for bwc / mixed cluster: adding new bytes to header would break older nodes (decoding the header), or I am missing something here?

Reply:
I didn't add new bytes. The `E` and `S` bytes as a prefix have always been there. Previously there was no need to store a field though because either the values were `E` and `S` or parsing failed. Now I'm storing an enum in preparation for different values being present (like `O` `P` for "OpenSearch Protobuf" or whatever). You can see in InboundDecoder it used to skip 6 bytes, whereas now it is reading 2 bytes and skipping 4 bytes.

Reply:
It's kind of confusing because this is where the 6 byte prefix is actually being read to get the message length:

https://github.com/opensearch-project/OpenSearch/blob/0138b6e3fd5d5d90c22b7b9f89dfdcfaa60af50e/server/src/main/java/org/opensearch/transport/InboundDecoder.java#L82

but when the rest of the header content is read that 6 byte prefix is just being skipped over:

https://github.com/opensearch-project/OpenSearch/blob/5e449764f521e42ff4d587d69b339064354b980f/server/src/main/java/org/opensearch/transport/InboundDecoder.java#L190

Reply:
Btw, a great place to learn about what this code actually does is to read the python implementation in https://github.com/opensearch-project/opensearch-sdk-py/tree/main/src/opensearch_sdk_py/transport :)